core/src/main/scala/akka/persistence/cassandra/journal/TagWriter.scala

```diff
@@ -0,0 +1,314 @@
+   */
+  private def calculateTagPidSequenceNr(pid: String, bucket: TimeBucket, tagPidSequenceNrs: Map[String, Future[SequenceNr]]): (Map[String, Future[SequenceNr]], Future[SequenceNr]) = {
+    val tagPidSequenceNr: Future[SequenceNr] = tagPidSequenceNrs.get(pid) match {
+      case None =>
+        log.debug("First time seeing this pId {} since startup", pid)
+        session.selectTagProgress(pid).map {
+          case None =>
+            log.debug("No previous progress for this tag/pid")
+            1
+          case Some(t @ TagProgress(_, _, pidTagSequenceNr)) =>
+            log.debug("Tag progress for new pId: {}", t)
+            pidTagSequenceNr + 1
+        }
+      case Some(n) =>
+        log.debug("Already cached. Previous tagPidSequenceNr: {}", n)
+        n.map(_ + 1)
+    }
+    (tagPidSequenceNrs + (pid -> tagPidSequenceNr), tagPidSequenceNr)
+  }
+
+  /**
+   * Events should be ordered by sequence nr per pid
+   */
+  private def write(events: Vector[Serialized], remainingBuffer: Vector[Serialized], existingTagSequenceNrs: Map[String, Future[Long]]): Unit = {
+    // Work out all the pidTagSequenceNrs
+    val (newTagSequenceNrs, withPidTagSeqNr) =
+      events.foldLeft((existingTagSequenceNrs, Vector.empty[Future[(Serialized, SequenceNr)]])) {
+        case ((accTagSequenceNrs, accEvents), nextEvent) =>
+          val (newSequenceNrs, sequenceNr) = calculateTagPidSequenceNr(nextEvent.persistenceId, nextEvent.timeBucket, accTagSequenceNrs)
+          (newSequenceNrs, accEvents :+ sequenceNr.map((nextEvent, _)))
+      }
+
+    // Create a summary of this write for logging and to be written to the
+    // tag write progress table
+    val withSummary = Future.sequence(withPidTagSeqNr)
+      .map((wpt: Seq[(Serialized, SequenceNr)]) => {
+        val writeSummary = wpt.foldLeft(Map.empty[PersistenceId, PidProgress])((acc, next) => {
+          val (event, tagPidSequenceNr) = next
+          acc.get(event.persistenceId) match {
+            case Some(PidProgress(from, to, _, _)) =>
+              if (event.sequenceNr <= to)
+                throw new IllegalStateException(s"Expected events to be ordered by seqNr. ${event.persistenceId} Events: ${events.map(e => (e.persistenceId, e.sequenceNr))}")
+              acc + (event.persistenceId -> PidProgress(from, event.sequenceNr, tagPidSequenceNr, event.timeUuid))
+            case None =>
+              acc + (event.persistenceId -> PidProgress(event.sequenceNr, event.sequenceNr, tagPidSequenceNr, event.timeUuid))
+          }
+        })
+        (wpt, writeSummary)
+      })
+
+    val done: Future[TagWriteFinished] = withSummary.flatMap {
+      case (writes, summary) =>
+        log.debug("Starting tag write. Summary: {}", summary)
+        session.writeBatch(writes).map(_ => TagWriteDone(summary))
+    }
+
+    // We never have concurrent writes for the same tag/node, go into writeInProgress state
+    context.become(writeInProgress(remainingBuffer, newTagSequenceNrs))
```

Comment:
how are failures handled here, can `tagPidSequenceNrs` have failed futures?

Reply:
We only have one Cassandra write in progress at any give time. That write will be flatMapped on the sequence number Future. If the overall future fails (could be for the sequenceNr read or the actual tag write) then we recert the tagSequenceNr map back to the state before the write removing any failed futures.

```
    case TagWriteFailed(t, events, previousSequenceNrs) =>
      log.error(t, "Writing tags has failed. This means that any eventsByTag query will be out of date. The write will be retried.")
      log.debug("Setting tag sequence nrs back to {}", previousSequenceNrs)
      // wait for the next poll as not to hammer Cassandra
      context.become(idle(events ++ buffer, previousSequenceNrs))
```

I test this here (with a terrible test name! should be handle sequence nr look up failing):

```
 "handle looking of bucket failing" in {
      val t = TestEx("No sequence nr for you")
      val (probe, ref) = setup(
        settings = defaultSettings.copy(maxBatchSize = 2, flushInterval = 50.milliseconds),
        sequenceNrLookup = Stream(Future.failed(t)) ++ Stream.continually(Future.successful(None))
      )
      val now = UUIDs.timeBased()
      val bucket = TimeBucket(now, bucketSize)

      val e1 = event("p1", 1L, "e-1", bucket)
      ref ! TagWrites(tagName, Vector(e1))
      probe.expectNoMessage(waitDuration)

      val e2 = event("p1", 2L, "e-2", bucket)
      ref ! TagWrites(tagName, Vector(e2))
      logProbe.expectMsgPF(waitDuration) {
        case Error(`t`, _, _, "Writing tags has failed. This means that any eventsByTag query will be out of date. The write will be retried.") =>
      }

      // next poll should work
      probe.expectMsg(Vector((e1, 1), (e2, 2)))
      probe.expectMsg(ProgressWrite("p1", 2, 2, e2.timeUuid))
    }
```

If you can think of a better way to keep track of the reading of tag sequence nrs on demand I am all ears, it felt a little dirty having a map of futures :)

Reply:
ok, that is good, that means that it will be retried which was what I wanted to be sure of